build_tools/github_actions/runner/instance_deleter/main.py

```diff
@@ -110,27 +115,110 @@ def _get_from_items(items: compute.Items, key: str):
   return next((item.value for item in items if item.key == key), None)
 
 
+def delete_instance_from_mig(mig_name: str, project: str, region: str,
+                             instance: compute.Instance):
+  try:
+    operation = migs_client.delete_instances(
+        instance_group_manager=mig_name,
+        project=project,
+        region=region,
+        # For some reason we can't just use a list of instance names and need to
+        # build this RhymingRythmicJavaClasses proto. Also, unlike all the other
+        # parameters, the instance has to be a fully-specified URL for the
+        # instance, not just its name.
+        region_instance_group_managers_delete_instances_request_resource=(
+            compute.RegionInstanceGroupManagersDeleteInstancesRequest(
+                instances=[instance.self_link])))
+  except (google.api_core.exceptions.Forbidden,
+          google.api_core.exceptions.Unauthorized,
+          google.api_core.exceptions.NotFound) as e:
+    print(e)
+    return flask.abort(
+        e.code, f"Error requesting that {mig_name} delete {instance.name}.")
+  except Exception as e:
+    # We'll call any other error here a server error.
+    print(e)
+    return flask.abort(
+        INTERNAL_SERVER_ERROR,
+        f"Error requesting that {mig_name} delete {instance.name}.")
+
+  try:
+    # This is actually an extended operation that you have to poll to get its
+    # status, but we just check the status once because it appears that errors
+    # always show up here and all we just want to return success in marking for
+    # deletion. We don't need to wait for the deletion to actually take place.
+    operation.result()
+  except google.api_core.exceptions.ClientError as e:
+    print(e)
+    # Unpack the actual usable error message
+    msg = (
+        f"Error requesting that {mig_name} delete {instance.name}:"
+        "\n" + "\n".join(
+            [f"{err.code}: {err.message}" for err in e.response.error.errors]))
+    print(msg)
+    # We're not actually totally sure whether this is a client or server error
+    # for the overall request, but let's call it a client error (the only client
+    # here is our VM instances, so I think we can be a bit loose).
+    return flask.abort(BAD_REQUEST, msg)
+
+  success_msg = f"{instance.name} has been marked for deletion by {mig_name}."
+  print(success_msg)
+  return success_msg
+
+
+def should_scale_down(mig_name: str, project: str, region: str):
+  start = time.time()
+  try:
+    print(f"Polling {mig_name} for stability")
+    while time.time() - start < STABILIZE_TIMEOUT_SECONDS:
```

Comment:
Well we certainly shouldn't just proceed as if it was as I'm currently doing! I'm not sure whether this should be an error or a success and return false. It doesn't really matter though because either way the self-deleter will halt. I'll return a 408, since that's for a timeout, though I'm not totally sure if it's the correct code...

Reply:
Edit: Actually returning a 504 after closer investigation